### PR TITLE
learn2016: Responsive page, on all domains

### PR DESCRIPTION
**Pull request by breville**
This is a big change which switches from brstrap to a React-based responsive system, adds hourofcode.com/learn and hourofcode.com/learn/robotics, and moves the filter list into JavaScript from HAML.

Our basic UI test matrix has gotten interesting...

**domain:** code.org, csedweek.org, hourofcode.com
**page:** /learn, /learn/robotics, /learn/legacy
**width:** 320-649, 650-819, 820-1023, 1024-1200

The TutorialExplorer component itself is shared between all the pages.  Its popup uses Bootstrap and so behaves a little differently between code.org/csedweek.org and hourofcode.com, partly because the former don't otherwise use Bootstrap, and partly because the latter uses an older version of Bootstrap 3.

### CODE REVIEW

File: apps/src/tutorialExplorer/responsive.jsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+function getItemWidthMobile(itemWidth, windowWidth) {
+  if (windowWidth < 600) {
+    return `${itemWidth}%`;
+  } else {
+    return "100%";
+  }
+}
+
+export { getContainerWidth, getItemWidth, getItemWidthMobile };
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/responsive.jsx]: Rather than passing around windowWidth as a prop, I might just have your responsive methods here calculate it (Means you don't have to worry about getting a mismatch between what a component thinks the window widht is and what these method thinks it is).

Downside is that it makes these methods less pure, as they now depend on browser, but that seems okay. I'd also add comments above these explaining (a) where these magic numbers come from and (b) what you're trying to accomplish.

File: apps/src/tutorialExplorer/filterGroup.jsx [hunk 2]
```diff
@@ -23,12 +24,15 @@ const FilterGroup = React.createClass({
...
     filterEntries: React.PropTypes.array.isRequired,
     selection: React.PropTypes.array.isRequired,
-    onUserInput: React.PropTypes.func.isRequired
+    onUserInput: React.PropTypes.func.isRequired,
+    windowWidth: React.PropTypes.number.isRequired
   },
 
   render() {
+    var filterGroupOuterStyle = {};
+    Object.assign(filterGroupOuterStyle, styles.filterGroupOuter, {float: "left", width: getItemWidthMobile(50, this.props.windowWidth)});
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/filterGroup.jsx]: note this can just be:

```
const filterGroupOuterStyle = Object.assign({}, styles.filterGroupOuter, { float: 'left', width: 'getItemWidthMobile(50, this.props.windowWidth);`
```

or if you want to be even more ES6y

```
const filterGroupOuterStyle = {
  ...styles.filterGroupOuter,
  float: 'left',
  width: getItemWidthMobile(50, this.props.windowWidth)
}
```

File: apps/src/tutorialExplorer/filterHeader.jsx [hunk 3]
```diff
@@ -49,6 +52,15 @@ const FilterHeader = React.createClass({
       <div style={styles.header}>
         {this.props.backButton && <BackButton/>}
 
+        {this.props.legacyLink && (
```
Comment (breville) [inline apps/src/tutorialExplorer/filterHeader.jsx:55]: We're only keeping this for about a week.

File: apps/src/tutorialExplorer/filterGroup.jsx [hunk 4]
```diff
@@ -28,7 +29,7 @@ const FilterGroup = React.createClass({
 
   render() {
     return (
-      <div style={styles.filterGroupOuter} className="col-mobile-50">
+      <div style={{...styles.filterGroupOuter, float: "left", width: getResponsiveValue({xs: 100, sm: 50, md: 100})}}>
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/filterGroup.jsx]: I would move this style object creation to live outside of the jsx, i.e.

``` jsx
const groupStyle = {
  ...styles.filterGroupOuter,
  float: "left",
  width: getResponsiveValue({xs: 100, sm: 50, md: 100})
};

return (
  <div style={groupStyle}>
```

File: apps/src/tutorialExplorer/filterHeader.jsx [hunk 5]
```diff
@@ -29,6 +31,7 @@ const styles = {
 const FilterHeader = React.createClass({
   propTypes: {
     backButton: React.PropTypes.bool,
+    legacyLink: React.PropTypes.string,
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/filterHeader.jsx:34]: What does it mean to be a legacy link?

File: apps/src/tutorialExplorer/filterHeader.jsx [hunk 6]
```diff
@@ -49,6 +52,15 @@ const FilterHeader = React.createClass({
       <div style={styles.header}>
         {this.props.backButton && <BackButton/>}
 
+        {this.props.legacyLink && (
+          <div>
+            <a href={this.props.legacyLink} style={{textDecoration: "underline", float: "right", paddingBottom: 6}}>
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/filterHeader.jsx:57]: can stick this style into your styles object? i.e. style={styles.legacyLink}

File: apps/src/tutorialExplorer/filterHeader.jsx (hunk 5 above)
Comment (breville) [reply apps/src/tutorialExplorer/filterHeader.jsx:34]: It's just a temporary link back to /learn/legacy and will be chopped in a week or less.

File: apps/src/tutorialExplorer/responsive.jsx [hunk 7]
```diff
@@ -0,0 +1,96 @@
+/* A very simple responsive layout system.
+ */
+
+/**
+ * Gets the container width.
+ * Returns either a number (e.g. 1170) or a string (e.g. "97%").
+ */
+
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/responsive.jsx]: nit: extra blank line

File: apps/src/tutorialExplorer/responsive.jsx [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  sm: 650,
+  xs: 0
+};
+
+/**
+ * Returns the window width that is the starting point for a width category.
+ *
+ * @param {string} id - "xs", "sm", "md", or "lg"
+ */
+
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/responsive.jsx]: nit: extra blank line

File: apps/src/tutorialExplorer/responsive.jsx [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+function getResponsiveWindowWidth(category) {
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/responsive.jsx]: I wonder if this would be slightly easier to follow if we did

```
function fitsResponsiveCategory(category, windowWidth) {
  return windowWidth >= responsiveWindowWidth[category];
}
```

then elsewhere we have

```
fitsResponsiveCategory("lg", windowWidth)
```

Alternatively, you could do:

```
function getResponsiveCategory(windowWidth) {
  if (windowWidth >= responsiveWindowWidth["lg"]) {
    return "lg"
 }
 // etc. for rest of the options
}
```

Then in getResponsiveValue()

```
const category = getResponsiveCategory(windowWidth());
if (category === "lg") {
 ...
}  else if (category === "sm") {
 ...
}
// etc
```

Any of these approaches (including yours) are probably fine, though I might like the last one best.

File: apps/src/tutorialExplorer/responsive.jsx [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+ * @param {number|string} values.sm - Value returned on small layout.
+ * @param {number|string} values.md - Value returned on medium layout.
+ * @param {number|string} values.lg - Value returned on large layout.
+
+ */
+
+function getResponsiveValue(values) {
+  const windowWidth = $(window).width();
+
+  var value;
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/responsive.jsx]: let instead of var

File: apps/src/tutorialExplorer/responsive.jsx [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+  }
+}
+
+/* Window widths that are the starting points for each width category. */
+const responsiveWindowWidth = {
+  lg: 1024,
+  md: 820,
+  sm: 650,
+  xs: 0
+};
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/responsive.jsx:29]: Any reason not to use full words instead of abbreviations?  Might also make sense to create an enum instead of using strings, i.e.:

```
// makeEnum comes from apps/src/utils
const ResponsiveSize = makeEnum('Large', 'Medium', 'Small', 'ExtraSmall');

const responsiveWindowWidth = {
  [ResponsiveSize.Large]: 1024,
  [ResponsiveSize.Medium]: 820
}
```

File: apps/src/tutorialExplorer/responsive.jsx [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+  if (value) {
+    if (typeof(value) === "number") {
+      return `${value}%`;
+    } else if (typeof(value) === "string") {
+      return value;
+    }
+  }
+}
+
+export { getResponsiveContainerWidth, getResponsiveWindowWidth, getResponsiveValue };
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/responsive.jsx]: instead of exporting everything at the end, you can just add `export` in front of each function

File: apps/src/tutorialExplorer/responsive.jsx [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+ *
+ * @param {Object} values - A set of values from which we want one.
+ * @param {number|string} values.xs - Value returned on extra-small layout.
+ * @param {number|string} values.sm - Value returned on small layout.
+ * @param {number|string} values.md - Value returned on medium layout.
+ * @param {number|string} values.lg - Value returned on large layout.
+
+ */
+
+function getResponsiveValue(values) {
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/responsive.jsx]: I really like how this works btw (being able to call getResponsiveValues with a set of different options based on sizing)

File: apps/src/tutorialExplorer/roboticsButton.jsx [hunk 14]
```diff
@@ -18,17 +19,27 @@ const styles = {
...
-      </a>
-    </div>
-    <div className="mobile-feature" style={styles.roboticsText}>
-      Got robots? <a href="/learn/robotics">Use these activities</a> and make a tangible Hour of Code for students of any age!
-    </div>
-  </div>
-);
+const RoboticsButton = React.createClass({
+  render() {
+
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/roboticsButton.jsx]: nit: extra blank line

File: apps/src/tutorialExplorer/roboticsButton.jsx [hunk 15]
```diff
@@ -18,17 +19,27 @@ const styles = {
...
+    return (
+      <div style={{float:"left"}}>
+        <div style={{display: getResponsiveValue({md: "block", xs: "none"})}}>
+          <a href="/learn/robotics">
+            <img src="/images/learn/robotics-link.png" style={styles.roboticsButtonImage}/>
+          </a>
+        </div>
+        <div style={roboticsTextStyle}>
+          Got robots? <a href="/learn/robotics">Use these activities</a> and make a tangible Hour of Code for students of any age!
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/roboticsButton.jsx:37]: Do we not need i18n?

File: apps/src/tutorialExplorer/tutorial.jsx [hunk 16]
```diff
@@ -50,27 +76,20 @@ const Tutorial = React.createClass({
...
-              <div style={styles.tutorialName}>
-                {this.props.item.name}
-              </div>
-              <div style={styles.tutorialSub}>
-                {getTutorialDetailString(this.props.item)}
-              </div>
-            </div>
+          <div style={styles.tutorialImageContainer}>
+            <img
+              src={this.props.item.image.replace("/images/", "/images/fill-480x360/").replace(".png", ".jpg")}
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/tutorial.jsx:84]: Why do we end up needing to change our image path?

File: apps/src/tutorialExplorer/tutorialDetail.jsx [hunk 17]
```diff
@@ -128,16 +136,13 @@ const TutorialDetail = React.createClass({
                 className="modal-body"
                 style={styles.tutorialDetailModalBody}
               >
-                <div className="col-50">
+                <div style={styles.tutorialDetailImageContainer} className="col-xs-12 col-sm-6">
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/tutorialDetail.jsx:139]: I see we're still using bootstrap for some of our responsiveness. How do we decide when to use/not use?

File: apps/src/tutorialExplorer/tutorialExplorer.js [hunk 18]
```diff
@@ -9,8 +9,10 @@ import Immutable from 'immutable';
 import FilterHeader from './filterHeader';
 import FilterSet from './filterSet';
 import TutorialSet from './tutorialSet';
+import { getResponsiveContainerWidth, getResponsiveWindowWidth } from './responsive';
 import _ from 'lodash';
 
+
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/tutorialExplorer.js]: nit: extra blank line

File: apps/src/tutorialExplorer/tutorialExplorer.js [hunk 19]
```diff
@@ -309,18 +306,101 @@ const TutorialExplorer = React.createClass({
...
+  };
+
+  if (robotics) {
+    for (var index = 0; index < filters.length; index++) {
+      var filterGroup = filters[index];
+      if (filterGroup.name === "activity_type") {
+        filterGroup.entries = [{name: "robotics", text: "Robotics"}];
+        filterGroup.display = false;
+      }
+    }
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/tutorialExplorer.js]: What you have is fine (though I'd use let instead of var). Could do this somewhat more concisely (tho this is probably me being somewhat nitty/personal preference):

``` js
filters.forEach(filterGroup => {
  if (filterGroup.name === "activity_type") {
    filterGroup.entries = [{name: "robotics", text: "Robotics"}];
    filterGroup.display = false;
 }
})
```

File: pegasus/sites.v3/hourofcode.com/public/learn/index.haml [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+  "og:video:type": ""
+---
+
+-# hourofcode.com uses an early version of Bootstrap 3 which didn't have this wider dialog.
+:css
+  @media (min-width: 992px)
+  {
+    .modal-lg {
+        width: 900px;
+  }
```
Comment (Bjvanminnen) [inline pegasus/sites.v3/hourofcode.com/public/learn/index.haml:22]: do we not need a close brace for .modal-lg?

File: pegasus/sites.v3/hourofcode.com/public/learn/index.haml [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+
+#fullwidth{style: "background-color: #2799a4; overflow: hidden"}
+  = view :header
+
+  .banner
+    .container{style: "color: white"}
+      .col-xs-12.col-sm-6{style: "text-align: center"}
+        %img{src: CDO.code_org_url("/images/learn/hour-of-code-logo-2016.png"), style: "margin-top: 30px; margin-bottom: 30px; max-width: 100%; padding-left: 5px; padding-right: 5px;"}
+        %br/
+        The Hour of Code is a global movement reaching tens of millions of students in 180+ countries. Anyone, anywhere can organize an Hour of Code event. One-hour tutorials are available in over 45 languages for all ages.
```
Comment (Bjvanminnen) [inline pegasus/sites.v3/hourofcode.com/public/learn/index.haml:32]: Do we need to care about i18n

File: pegasus/sites.v3/hourofcode.com/public/learn/index.haml [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+    var studentImage = "#{CDO.code_org_url('/images/learn/student')}" + studentImageIndex + '.png';
+
+    var img = $('<img>');
+    img.attr('src', studentImage);
+    img.attr('style', 'opacity: 0');
+    img.appendTo('#studentimage');
+    img.on('load', function() {
+      img.fadeTo('normal', 1);
+    });
+  });
```
Comment (Bjvanminnen) [inline pegasus/sites.v3/hourofcode.com/public/learn/index.haml:86]: This is starting to look like more code than I'd like to see in a haml file.

We came up with a cleaner approach for getting data from dashboard/haml to JS in https://github.com/code-dot-org/code-dot-org/pull/11322. Probably worth doing something like that here, but it might make sense to do that as a separate PR to make things less confusing. Happy to help you out to get that working.

File: pegasus/sites.v3/hourofcode.com/public/learn/index.haml [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+      legacyLink: "#{CDO.code_org_url('/learn/legacy')}"
+      }
+    );
+    tutorialExplorerManager.renderToElement(document.getElementById('tutorials'));
+
+    // Pick a random student image (1-2 inclusive) and fade it in.
+    var studentImageIndex = Math.floor(Math.random() * 2) + 1;
+    var studentImage = "#{CDO.code_org_url('/images/learn/student')}" + studentImageIndex + '.png';
+
+    var img = $('<img>');
```
Comment (Bjvanminnen) [inline pegasus/sites.v3/hourofcode.com/public/learn/index.haml:79]: bunch of vars that should be lets.

File: pegasus/sites.v3/hourofcode.com/public/learn/robotics.haml [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+  "og:video:height": ""
+  "og:video:type": ""
+---
+
+-# hourofcode.com uses an early version of Bootstrap 3 which didn't have this wider dialog.
+:css
+  @media (min-width: 992px)
+  {
+    .modal-lg {
+        width: 900px;
```
Comment (Bjvanminnen) [inline pegasus/sites.v3/hourofcode.com/public/learn/robotics.haml]: same comment about close brace. im surprised you dont get errors about this when you load the page?

File: pegasus/sites.v3/hourofcode.com/public/learn/robotics.haml [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+        %br/
+        %i{style: "font-family: 'Gotham 5r', sans-serif; font-size: 56px"}
+          ROBOTICS
+        %br/
+        %br/
+        Got robots? Use them with the activities below and make a tangible Hour of Code for students of any age!
+      #studentimage.col-xs-12.col-sm-6.hidden-xs
+        %img{src: CDO.code_org_url("/images/learn/robotics.png"), style: "text-align: center"}
+
+-# Not ideal but we can pull the undigested file from /blockly.
```
Comment (Bjvanminnen) [inline pegasus/sites.v3/hourofcode.com/public/learn/robotics.haml:40]: Im wondering if we need to figure out a better solution for this before HOC

**Comment by Bjvanminnen** [comment]
Sorry for the excess of comments. Feel free to ignore many of them as you wish. I do quite like this general approach for responsiveness :)

It would be nice to cleanup the communication between dashboard/apps here in a future PR. I bet we can reduce our usage of globals :)

**Comment by tanyaparker** [comment]
Quick tests on Chrome, Safari, FF -- lgtm. It's a little slow on the change if you're resizing your browser, but I don't think people do this often and we don't need to optimize this as much as loading the pictures.

File: apps/src/tutorialExplorer/responsive.jsx (hunk 11 above)
Comment (breville) [reply apps/src/tutorialExplorer/responsive.jsx:29]: Only that the abbreviations match what Bootstrap uses (though our resolutions don't anymore).

File: apps/src/tutorialExplorer/responsive.jsx (hunk 13 above)
Comment (breville) [reply apps/src/tutorialExplorer/responsive.jsx]: Yeah, this is basically how you declare grid widths with Bootstrap classes, but with code.

File: apps/src/tutorialExplorer/roboticsButton.jsx (hunk 15 above)
Comment (breville) [reply apps/src/tutorialExplorer/roboticsButton.jsx:37]: Soon we'll loc everything in a fresh PR.

File: apps/src/tutorialExplorer/tutorial.jsx (hunk 16 above)
Comment (breville) [reply apps/src/tutorialExplorer/tutorial.jsx:84]: The gsheet lists .png files but by changing them to .jpg Pegasus will reformat them before serving, and we save a bunch of bandwidth.

File: apps/src/tutorialExplorer/tutorialDetail.jsx (hunk 17 above)
Comment (breville) [reply apps/src/tutorialExplorer/tutorialDetail.jsx:139]: The mash-up is certainly not ideal, but I wanted tutorialExplorer to take over responsive layout for all its regular UI.  However, we're using Bootstrap's modal dialog, so I'm using entirely Bootstrap for that, which is contained in tutorialDetail.jsx only.

File: pegasus/sites.v3/hourofcode.com/public/learn/index.haml (hunk 23 above)
Comment (breville) [reply pegasus/sites.v3/hourofcode.com/public/learn/index.haml:79]: Do we get `let` in our full browser matrix?  (This code doesn't go through pipeline since it's inline in HAML.)

File: pegasus/sites.v3/hourofcode.com/public/learn/robotics.haml (hunk 25 above)
Comment (breville) [reply pegasus/sites.v3/hourofcode.com/public/learn/robotics.haml:40]: Definitely interested in this comment.  Scaling risk?

File: pegasus/sites.v3/hourofcode.com/public/learn/index.haml (hunk 23 above)
Comment (Bjvanminnen) [reply pegasus/sites.v3/hourofcode.com/public/learn/index.haml:79]: Good point. No we don't. Another reason to move this out of haml ;-)

File: apps/src/tutorialExplorer/responsive.jsx (hunk 9 above)
Comment (breville) [reply apps/src/tutorialExplorer/responsive.jsx]: Good idea.  Taking a slightly different approach so that we don't need to pass current window width to the function, since the rest of responsive.jsx now takes care of retrieving current window width directly rather than requiring it to be passed in.
